plugins/inputs/nfsclient/nfsclient.go

```diff
@@ -0,0 +1,421 @@
+  # for different mountpoints:  use mulitple [[input.nfsclient]] stanzas,
+  # with their own lists.  See "include_mounts" above, and be careful of
+  # duplicate metrics.
+  include_operations = []
+  exclude_operations = []
+`
+
+func (n *NFSClient) SampleConfig() string {
+	return sampleConfig
+}
+
+func (n *NFSClient) Description() string {
+	return "Read per-mount NFS metrics from /proc/self/mountstats"
+}
+
+var eventsFields = []string{
+	"inoderevalidates",
+	"dentryrevalidates",
+	"datainvalidates",
+	"attrinvalidates",
+	"vfsopen",
+	"vfslookup",
+	"vfsaccess",
+	"vfsupdatepage",
+	"vfsreadpage",
+	"vfsreadpages",
+	"vfswritepage",
+	"vfswritepages",
+	"vfsgetdents",
+	"vfssetattr",
+	"vfsflush",
+	"vfsfsync",
+	"vfslock",
+	"vfsrelease",
+	"congestionwait",
+	"setattrtrunc",
+	"extendwrite",
+	"sillyrenames",
+	"shortreads",
+	"shortwrites",
+	"delay",
+	"pnfsreads",
+	"pnfswrites",
+}
+
+var bytesFields = []string{
+	"normalreadbytes",
+	"normalwritebytes",
+	"directreadbytes",
+	"directwritebytes",
+	"serverreadbytes",
+	"serverwritebytes",
+	"readpages",
+	"writepages",
+}
+
+var xprtudpFields = []string{
+	"bind_count",
+	"rpcsends",
+	"rpcreceives",
+	"badxids",
+	"inflightsends",
+	"backlogutil",
+}
+
+var xprttcpFields = []string{
+	"bind_count",
+	"connect_count",
+	"connect_time",
+	"idle_time",
+	"rpcsends",
+	"rpcreceives",
+	"badxids",
+	"inflightsends",
+	"backlogutil",
+}
+
+var nfs3Fields = []string{
+	"NULL",
+	"GETATTR",
+	"SETATTR",
+	"LOOKUP",
+	"ACCESS",
+	"READLINK",
+	"READ",
+	"WRITE",
+	"CREATE",
+	"MKDIR",
+	"SYMLINK",
+	"MKNOD",
+	"REMOVE",
+	"RMDIR",
+	"RENAME",
+	"LINK",
+	"READDIR",
+	"READDIRPLUS",
+	"FSSTAT",
+	"FSINFO",
+	"PATHCONF",
+	"COMMIT",
+}
+
+var nfs4Fields = []string{
+	"NULL",
+	"READ",
+	"WRITE",
+	"COMMIT",
+	"OPEN",
+	"OPEN_CONFIRM",
+	"OPEN_NOATTR",
+	"OPEN_DOWNGRADE",
+	"CLOSE",
+	"SETATTR",
+	"FSINFO",
+	"RENEW",
+	"SETCLIENTID",
+	"SETCLIENTID_CONFIRM",
+	"LOCK",
+	"LOCKT",
+	"LOCKU",
+	"ACCESS",
+	"GETATTR",
+	"LOOKUP",
+	"LOOKUP_ROOT",
+	"REMOVE",
+	"RENAME",
+	"LINK",
+	"SYMLINK",
+	"CREATE",
+	"PATHCONF",
+	"STATFS",
+	"READLINK",
+	"READDIR",
+	"SERVER_CAPS",
+	"DELEGRETURN",
+	"GETACL",
+	"SETACL",
+	"FS_LOCATIONS",
+	"RELEASE_LOCKOWNER",
+	"SECINFO",
+	"FSID_PRESENT",
+	"EXCHANGE_ID",
+	"CREATE_SESSION",
+	"DESTROY_SESSION",
+	"SEQUENCE",
+	"GET_LEASE_TIME",
+	"RECLAIM_COMPLETE",
+	"LAYOUTGET",
+	"GETDEVICEINFO",
+	"LAYOUTCOMMIT",
+	"LAYOUTRETURN",
+	"SECINFO_NO_NAME",
+	"TEST_STATEID",
+	"FREE_STATEID",
+	"GETDEVICELIST",
+	"BIND_CONN_TO_SESSION",
+	"DESTROY_CLIENTID",
+	"SEEK",
+	"ALLOCATE",
+	"DEALLOCATE",
+	"LAYOUTSTATS",
+	"CLONE",
+}
+
+var nfsopFields = []string{
+	"ops",
+	"trans",
+	"timeouts",
+	"bytes_sent",
+	"bytes_recv",
+	"queue_time",
+	"response_time",
+	"total_time",
+	"errors",
+}
+
+func convert(line []string) []int64 {
+	var nline []int64
+	for _, l := range line[1:] {
+		f, _ := strconv.ParseInt(l, 10, 64)
```

Comment:
We _really_ should check the error here!

Reply:
The basic logic relies on dropping non-number values, for better or worse.  For example, given this line:
```
OPEN: 19740 19740 0 4737600 7343280 505 3449 4172
```
An array is returned, and later used:
```
19740 19740 0 4737600 7343280 505 3449 4172
```

But all lines are actually checked, even if nothing is actually used.  This arguably needs to get refactored, but that will be part of a larger rewrite.  For example the line below parses out as follows, and returns the array at the very end:

```
RPC iostats version: 1.0 p/v: 100003/4 (nfs)
ErrRange: [RPC iostats version: 1.0 p/v: 100003/4 (nfs)] [iostats] -> [0]
ErrRange: [RPC iostats version: 1.0 p/v: 100003/4 (nfs)] [version:] -> [0]
ErrRange: [RPC iostats version: 1.0 p/v: 100003/4 (nfs)] [1.0] -> [0]
ErrRange: [RPC iostats version: 1.0 p/v: 100003/4 (nfs)] [p/v:] -> [0]
ErrRange: [RPC iostats version: 1.0 p/v: 100003/4 (nfs)] [100003/4] -> [0]
ErrRange: [RPC iostats version: 1.0 p/v: 100003/4 (nfs)] [(nfs)] -> [0]
[0 0 0 0 0 0]
```

The weird values we'd get out are `0` when the number can't convert at all, or  `±MAXINT` in the case over/under-flow values. 
Printing an error on every line is a mistake, because that happens even on a line of "valid" input data.  Printing an error when out of range might be useful though, since it will 1) the user data is getting messed with unexpectedly, and  2)  indicate a possible problem with the upstream datasource (_i.e._ mountstats)

Reply:
Will comment below. Please mark as resolved.